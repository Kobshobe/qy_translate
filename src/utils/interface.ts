import {Ref} from 'vue'

export class Find {
    text: string
    result: ITransResult | undefined
    isCollected: boolean
    tid: number | null
    constructor(text:string) {
        this.text = text
        this.result = undefined
        this.isCollected = false
        this.tid = null
    }
}

export interface ITokenInfo {
    token: string
    saveTime: number
    liveTime: number
    openid: string
}

export interface ITokenInfoFromCloud {
    token: string
    openid: string
    liveTime: number
}

export interface ICollectResultMsg {
    statusMsg: string
    isCollected: boolean
    tid: number
}

export interface ITransInfo {
    text: string
    from: string
    to: string
    type: string
    mode: string
}


export interface IRequestResult {
    errMsg: string
    data?: any
    status?: number
    toastMsg?: string
    serveToastMsg?: string
}

export interface ITransResult {
    text: string
    resultFrom: string
    resultTo: string
    pronunciation: string
    data: any
}

export interface ITransResultFromApi {
    sentences: any[]
    dict?: IDictItem[]
    src: string
    alternative_translations: any[]
    confidence: number
    spell: any
    Id_result: any
    synsets: any[]
    examples?: any[]
}

export interface IDictItem {
    pos: string
    terms: string[]
    entry: string[]
    base_from: string
    pos_enum: number
}

export interface IMarkInfo {
    canReduce: boolean
    newMarksList: number[][]
}

export interface IOptionPageOpenParma {
    tab: 'login' | ''
}

export interface ILangInfo {
    main: string,
    secord: string
}

export interface IAnalyticEvent {
    name: string
    params: any

}

export interface IEventList {
    name: string
    client: string
    clientVersion: string
    eventList: any[]
}

export interface eventItem {
    name: string
    type: "wrap" | "single"
    time: string
    duration: number
}

export interface ITranslateMsg {
    text: string,
    type: 'sub' | 'edit_icon' | 'select' | 'changeLang' | 'edit_enter' | 'test'|'exchange'
    from?: string
    to?: string
    findStatus?: 'loading'|undefined
}

export interface ITranslatorHook {
    mode: 'resultOnly'|'popup'|'pdf'
    status: 'result' | 'editing'
    editingText: string,
    show: boolean,
    find: Find,
    findStatus: 'none' | 'ok' | 'loading' | 'reLoad',
    fromIframe: any,
    toIframe: any,
    subIfram: any,
    dialogMsg: {
        show: boolean
        contentText: string
        confirmText: string
        cancelText: string
        confirmAction: any
        showDialog(contentText: string, confirmText: string, cancelText: string, confirmAction: any): void
    }
    toast: {
        show: boolean
        msg: string
        closeTimer: any
        showToast({ msg, duration }: { msg: string, duration?: number }): void
    }
    marksList: any[]
    canReduceMark: boolean
    subTranslator: {
        status: 'loading' | 'hide' | 'result' | 'showGate'
        selectText: string
        selectRange: number[]
        top: number
        left: number
        isLoading: boolean
        resultData: ITransResult | undefined
        marksStr: string,
        translate(): void
        afterMark(canReduce: boolean): void
        bookMark(): void
        mark(): void
        init(): void
    }
    options: {
        isShow: boolean
        from: string|undefined
        to: string|undefined
        close():void
        openOptionsPage():void
        show() :void
        exchange() :void
    }
    configInfo: {
        isTreadWord: boolean
        info: IConfigInfo|{}
        getTreadWord() :void
        changeTreadWord() :void
    }
    usePort({ name, msg, onMsgHandle }: { name: string, msg: any, onMsgHandle: Function }): void
    handleWebErr(msg: IRequestResult): void
    getMarkHtml(): void
    updateMark({ success, fail, info }: { success: Function, fail: Function, info: {marks: string, tid: number } }): void
    reduceCollect(): void
    collect({ success, fail }: { success?: Function, fail?: Function }): void
    translateText({ text, from, to, type, findStatus }: ITranslateMsg): void
    translateFromEdit(e: any): void
    getTTS(audioType: string, id: string): void
    clear(): void
    toEdit(): void
    copyResult(): void
    eventToAnalytic(eventData: any): void
    toAnalytics(event: IAnalyticEvent) :void
}

export interface IQrLoginParams {
    qrUrl: Ref<string>
    loginStatus: Ref<'loginOk' | 'none' | 'scanQr' | 'loadingQr' | 'invalidQr' | 'loadQrFail'>
}

export interface IConfigInfo {
    isTreadWord?: boolean
}