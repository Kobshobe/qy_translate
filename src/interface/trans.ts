import {Ref} from 'vue'
import { Context,IDialogMsg,IToastMsg } from '@/api/context'

export declare type ITransMode = 'popup'|'contentInject'|'pdf'

export declare type ITransStatus = 'editing'|'result'|'none'

export declare type ITransEngine = ''|'ggTrans__common'|'bdTrans__common'|'bdDM__finance'|
'bdDM__electronics'|'bdDM__mechanics'|'bdDM__medicine'|'bdDM__novel'

export declare type ITransType = 'sub'|'popup_icon'|'select'|'changeLang'|'popup_enter'|'popup_shift_enter'|'test'|
'exchange'|'popup_paste'|'menu'|'changeEngine'

export declare type IPortName = 'translate'|'collect'|'reduceCollect'|'updateMark'|'tts'|
'openOptionsPage'|'analytic'|'applyBDDM'

export declare type ITransFindStatus = 'none'|'ok'|'popLoading'|'reLoading'|'willOK'|'editLoading'

export interface IBaseHook {
    mode: ITransMode
    isResultInit: boolean
    status:ITransStatus
    findStatus: ITransFindStatus
    transID: number
    isHold: boolean
    lockLang: boolean
    bridge: any
    C: IAllStorage
    init(mode:ITransMode, status:ITransStatus) :void
    getConf() :void
    changeTreadWord() :void
    exchangeLang():void
    changeLang(isExchange:boolean):void
    eventToAnalytic(eventData:any) :void
    usePort({ name, context, onMsgHandle}:IPortHandler): void
    setNoneStatus(changeID?:boolean):void
    setHold() :void
    openOptionsPage(type:string):void
    dialog: {
        show: boolean
        message: string
        confirmText: string|undefined
        cancelText: string|undefined
        confirmAction: any
        showDialog(dialogMsg:IDialogMsg): void
    }
    toast: {
        show: boolean
        msg: string
        closeTimer: any
        showToast(toastMsg:IToastMsg): void
    },
    tips: {
        messages: string[]
        message: any
    }
    E: IEditHook
    T:ITranslatorHook
}

export interface IEditHook {
    base: IBaseHook
    editingText: string
    lastFindText: string
    clear() :void
    getLastFindText() :void
    pasteAndTrans() :void
    trans(type?:ITransType) :void
    enterTrans(e:any) :void
}

export interface ITranslatorHook {
    base: IBaseHook
    lastFindText: string
    show: boolean
    find: Find
    fromIframe: any
    toIframe: any
    subIfram: any
    marksList: any[]
    canReduceMark: boolean
    subTranslator: {
        status: 'loading'|'hide'|'result'|'showGate'
        selectText: string
        selectRange: number[]
        top: number
        left: number
        isLoading: boolean
        resultData: ITransResult | undefined
        marksStr: string,
        translate(): void
        afterMark(canReduce: boolean): void
        bookMark(): void
        mark(): void
        init(): void
    }
    options: {
        isShow: boolean
        from: string|undefined
        to: string|undefined
        engine: ITransEngine
        getData():void
        close():void
        show() :void
        exchange() :void
        changeEngine():void
        setLang():void
    }
    usePort({ name, context, onMsgHandle}:IPortHandler): void
    handleWebErr(msg: Context): void
    getMarkHtml(): string
    updateMark({ success, fail, info }: { success: Function, fail: Function, info: {marks: string, tid: number } }): void
    reduceCollect(): void
    collect({ success, fail }: { success?: Function, fail?: Function }): void
    trans(info: ITransMsg): void
    translateFromEdit(): void
    getTTS(audioType: string, id: string): void
    toEdit(): void
    copyResult(): void
    eventToAnalytic(eventData: any): void
    applyBDDM() :void
    setResultPostion():void
}

export interface IAllStorage {
    mainLang?: string
    secondLang?: string
    fromLang?: string
    toLang?:string
    isTreadWord?: boolean
    tokenInfo?: ITokenInfo
    mode?: 'simple'| 'profession'
    proMode?: 'float'|'edge'
    optionPageOpenParmas?: IOptionPageOpenParma
    menuTrans?: boolean
    transEngine?: ITransEngine
    showProun?: boolean
    keyDownTrans?: 'Enter'|'Shift+Enter'
}

export class Find {
    text: string
    result: ITransResult | undefined
    isCollected: boolean
    tid: number | null
    constructor(text:string) {
        this.text = text
        this.result = undefined
        this.isCollected = false
        this.tid = null
    }
}

export interface ITokenInfo {
    token: string
    saveTime: number
    liveTime: number
    openid: string
}

export interface ITokenInfoFromCloud {
    token: string
    openid: string
    liveTime: number
}

export interface ICollectResultMsg {
    statusMsg: string
    isCollected: boolean
    tid: number
}

export interface IWrapTransInfo {
    readonly text: string
    from?: string
    to?: string
    readonly type: string
    readonly mode: string
    engine?: string
    isDetectedLang?: boolean
    fromCode?: string
    toCode?: string
    fromCp?: string
    toCp?: string
    sFrom?: string
    sTo?: string
    cost?: number
    extraMsg?: Map<string, any>
}

export interface ITransResult {
    text: string
    resultFrom: string
    resultTo: string
    engine: string
    sPronunciation?: string
    tPronunciation?: string
    data?: any
    domain?: string
    examples?: ITransExample[]
    dict?: ITransDict[]
}

export interface ITransExample {
    text: string
    trans?: string
}

export interface ITransDict {
    pos: string
    trans: string
}

export interface IGTransResult {
    sentences: any[]
    dict?: IDictItem[]
    src: string
    alternative_translations: any[]
    confidence: number
    spell: any
    Id_result: any
    synsets: any[]
    examples?: any[]
}

export interface IDictItem {
    pos: string
    terms: string[]
    entry: string[]
    base_from: string
    pos_enum: number
}

export interface IMarkInfo {
    canReduce: boolean
    newMarksList: number[][]
}

export interface IOptionPageOpenParma {
    tab: 'login' | ''
    action?: string
}

export interface ILangInfo {
    main: string,
    secord: string
}

export interface IAnalyticEvent {
    name: string
    params: any
}

export interface eventItem {
    name: string
    type: "wrap" | "single"
    time: string
    duration: number
}

export interface ITransMsg {
    text: string,
    type: ITransType
    from?: string
    to?: string
    findStatus: ITransFindStatus
    engine?: string
    id?: number
}

export interface IPortHandler {
    name: IPortName
    context: Context
    onMsgHandle?(context:Context) :void
}

export interface IQrLoginParams {
    qrUrl: Ref<string>
    loginStatus: Ref<'loginOk' | 'none' | 'scanQr' | 'loadingQr' | 'invalidQr' | 'loadQrFail'>
}

export interface IClientInfo {
    c: string
    os: string
    l: string
    cv: string
    st: string
}

export interface IBDDMTransResult {
    result: string[]
    from: string
}
