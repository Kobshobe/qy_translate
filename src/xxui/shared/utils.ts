// import { isArray, isObject, isString } from '@vue/shared'

// export {
//   isArray,
//   isFunction,
//   isObject,
//   isString,
//   isDate,
//   isPromise,
//   isSymbol,
// } from '@vue/shared'
// export { isVNode } from 'vue'

// export const isUndefined = (val: any): val is undefined => val === undefined
// export const isBoolean = (val: any): val is boolean => typeof val === 'boolean'
// export const isNumber = (val: any): val is number => typeof val === 'number'

// export const isEmpty = (val: unknown) =>
//   (!val && val !== 0) ||
//   (isArray(val) && val.length === 0) ||
//   (isObject(val) && !Object.keys(val).length)

// export const isElement = (e: unknown): e is Element => {
//   if (typeof Element === 'undefined') return false
//   return e instanceof Element
// }

// export const isStringNumber = (val: string): boolean => {
//   if (!isString(val)) {
//     return false
//   }
//   return !Number.isNaN(Number(val))
// }

export const isUndefinedOrNull = (val:any) :boolean => {
  if (val === null || val === undefined) {
    return true
  } else {
    return false
  }
}

export function findIndex(array: any[], value: any) {
  return array.findIndex(item => item === value)
}