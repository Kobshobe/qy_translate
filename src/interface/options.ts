// import {TTS} from './translator'

import { IAllStorage } from "./trans";
import { Context } from "@/api/context";

export interface IOptionBaseHook {
  OP: IConfHook
  coll: ICollHook
  user: {
    isLogin: boolean
    isShowLogin: boolean
    logout() :void
  }
}

export interface IConfHook {
  base: IOptionBaseHook
  conf: {
      C:IAllStorage,
      changeTransEngine() :void
      changeTreadWord() :void
      changeMode() :void
      changeMenuTrans() :void
      changeMainLang() :void
      changeSecondLang() :void
      changeShowProun() :void
      changeShowProun() :void
      changeKeyDownTrans() :void
  }
  getOpenParams():void
  init():void
}

export interface ICollHook {
  base: IOptionBaseHook
  tid: undefined | number
  name: string
  page: number,
  status: "none" | "edit"
  loadStatus: "none" | "loading" | "loaded" | "loadFail" | "loadingMore" | "noMore"|'empty'|'needLogin'
  showMarks: boolean
  showParaphrase: boolean
  collList: ICollection[]
  phraseList: IPhraseInfo[]
  pageSize: number
  tts: any
  playTTSInfo: IPlayTTSInfo
  selected: Set<number>
  setPlayMode(): void
  setAtSelected(tid: number) :void
  setStatus(s: string) :void
  moveToColl(e: any) :void
  deleteSelected(): void
  playList(): void
  stopPlayTTS(): void
  playTTS(index: number, isPlayList?: boolean) :void
  getCollList(): void
  getPhraseList(collId: number|undefined, page: number):void
  setShowMarks(): void
  setShowParaphrase(): void
  back() :void
  handleRespMsg(c:Context) :void
  init() :void
  initColl() :void
  collItem: {
    isShowDialog: boolean
    collName: string
    oldCollName: string
    collTid?: number
    deleted: number[]
    create() :void
    rename() :void
    delete(tid:number) :void
    showDialog(name?:string, tid?:number) :void
    createOrRename() :void
  }
}


export interface ITokenInfo {
  token: string
  saveTime: Date
}

export interface IPlayTTSInfo {
  nowPlayIndex: number
  isPlayList: boolean
  playMode: 'single' | 'cycle' | 'order'
  isChoice: boolean
  init(): void
}

export interface IPhraseInfo {
  text: string
  tid: number
  translation: string
  marks: string
  resultFrom: string
  resultTo: string
}

export interface ICollection {
  tid: number
  name: string
  UpdatedAt: number
}

export interface ICollPhraseInfo {
  text: string
  translation: string
  marks: string
  resultFrom: string
  resultTo: string
}

export interface ITransResult {
  text: string
  resultFrom: string
  resultTo: string
  pronunciation: string
  data: any
}

export interface IIndexHook {
  loadStatuse: string
  showAddCollInput: boolean,
  phraseList: any[],
  collList: any
  getCollectionList() :void
  addColl(name:string):void
  toFind() :void
  toCollectionManagerPage() :void
  toTips() :void
}

export interface IFindHook {
  mode: string
  text: string
  tid: undefined|number,
  result: undefined|ITransResult,
  tts: any,
  clear() :void
  find() :void
  copyResult() :void
  collect(): void
  playTTS(type:'from'|'to') :void
}