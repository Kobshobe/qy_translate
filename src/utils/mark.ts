import {IMarkInfo} from '@/interface/trans'

export function newMarkManager(mark:number[], marksList:number[][]) :IMarkInfo {
    const newMarksList = []
    let canReduce = false

    marksList.forEach((range) => {
        if (range[1] < mark[0] || range[0] > mark[1]) {
            newMarksList.push(range)
        } else if (range[0] <= mark[0] && range[1] >= mark[1]) {
            canReduce = true
            if(range[0] < mark[0]) {
                newMarksList.push([range[0], mark[0]])
            }
            if(range[1] > mark[1]) {
                newMarksList.push([mark[1], range[1]])
            }
        } else {
            mark[0] = Math.min(range[0], mark[0])
            mark[1] = Math.max(range[1], mark[1])
        }
    })

    if(!canReduce) {
        newMarksList.push(mark)
    }
    newMarksList.sort((a, b) :number => {
        return a[0] - b[0]
    })

    return {
        canReduce,
        newMarksList
    }
}

export function getMarkHtml(marksList:number[][], text:string) :string {
    let markText = ''
    let point = 0;
    if (marksList.length > 0) {
      marksList.forEach((range: Array<number>) => {
        if (range[0] >= point) {
          markText += text.slice(point, range[0]);
          markText += `<span class="mark-text">${text.slice(
            range[0],
            range[1]
          )}</span>`;
          point = range[1];
        }
      });
    }
    markText += text.slice(point);
    return markText
}

export function getMarkHtmlFromStr(marksStr:string, text:string) {
    if(!marksStr) return text
    let marksList:number[][];
    try {
       marksList = JSON.parse(marksStr)
       return getMarkHtml(marksList, text)
    } catch {
        return text
    }
}