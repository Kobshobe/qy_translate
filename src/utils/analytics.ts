import { IEvent } from './interface'
import { googleAnalytic, clientVersion, client, Mode, os } from '@/config'
import { getClientId } from './chromeApi'


export async function eventToGoogle(event: IEvent) {
    const id = await getClientId()
    Object.assign(event.params, {cv:clientVersion, c:client, st: navigator.userAgent, mode: Mode, os})
    fetch(
        `https://www.google-analytics.com/mp/collect?measurement_id=${googleAnalytic.measurementId}&api_secret=${googleAnalytic.apiSecret}`,
        {
            method: "POST",
            body: JSON.stringify({
                client_id: id,
                user_id: id,
                events: [
                    {
                        name: event.name,
                        params: event.params
                    },
                ],
            }),
        }
    );
}

export function eventToAnalytic({data}:{data:any}) {
    eventToGoogle(data)
}