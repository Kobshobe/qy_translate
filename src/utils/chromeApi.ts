import { Mode } from '@/config'
import { ITokenInfo, ITokenInfoFromCloud, IAllStorage } from '@/interface/trans'
import { v4 } from "uuid";
import { eventToGoogle } from './analytics'
import { languages } from '@/translator/trans_base'
import { clientVersion } from '@/config'

export async function getTransConf(): Promise<IAllStorage> {
    const conf: IAllStorage = await getFromeStorage([
        'isTreadWord', 'fromLang', 'toLang', 'mode', 'transEngine', 'showProun', 'keyDownTrans',
        'mainLang', 'secondLang',
    ])
    conf.isTreadWord = dealTreadWord(conf.isTreadWord);
    conf.menuTrans = dealTreadWord(conf.menuTrans);
    conf.showProun = dealTreadWord(conf.showProun)
    conf.fromLang || (conf.fromLang = 'auto');
    conf.toLang || (conf.toLang = '__auto__');
    conf.mode || (conf.mode = 'simple');
    conf.transEngine || (conf.transEngine = 'bdTrans__common');
    conf.keyDownTrans || (conf.keyDownTrans = 'Enter');
    if (!conf.mainLang) {
        [conf.mainLang, conf.secondLang] = setLang();
        eventToGoogle({
            name: 'errConfSetLang',
            params: {}
        })
    }

    return conf
}

function dealTreadWord(isTreadWord: any): boolean {
    if (isTreadWord === false) {
        return false
    } else {
        return true
    }
}

export async function getFromeStorage(field: string[]): Promise<IAllStorage> {
    return new Promise<IAllStorage>((resolve) => {
        chrome.storage.sync.get(field, (result) => {
            resolve(result)
        })
    })
}

export async function getClientId(): Promise<string> {
    return new Promise<string>((resolve) => {
        chrome.storage.sync.get(['tokenInfo', 'uuid'], (result: any) => {
            const tokenInfo: ITokenInfo | null = result.tokenInfo
            if (tokenInfo && tokenInfo.openid) {
                resolve("op:" + tokenInfo.openid)
            } else {
                if (result.uuid) {
                    resolve("uuid:" + result.uuid)
                } else {
                    const uuid = v4()
                    chrome.storage.sync.set({ uuid })
                    resolve("nuuid:" + uuid)
                }
            }
        })
    })
}

export function getTokenFromStorage(getCheckToken = true): Promise<string> {

    return new Promise<string>((resolve, reject) => {
        chrome.storage.sync.get(['tokenInfo'], (result: any) => {
            const tokenInfo: ITokenInfo | null = result.tokenInfo
            const check = checkToken(tokenInfo)
            if (check === '__needRelogin__' || check === '__needLogin__') {
                reject(check)
            } else {
                resolve(check)
            }
        })
    }).then(res => { return res }).catch(res => { return res })
}

function checkToken(tokenInfo: ITokenInfo | null): string {
    if (tokenInfo) {
        const now = new Date().getTime()
        if (now - tokenInfo.saveTime + 1800 > tokenInfo.liveTime) {
            return '__needRelogin__'
        }
        return tokenInfo.token
    } else {
        return '__needLogin__'
    }
}

export function saveTokenInfo(info: ITokenInfoFromCloud, callBack: Function) {
    const now = new Date()
    const tokenInfo: ITokenInfo = {
        saveTime: now.getTime(),
        token: info.token,
        liveTime: info.liveTime * 3600 * 1000,
        openid: info.openid
    }
    chrome.storage.sync.set({ tokenInfo: tokenInfo }, function () {
        callBack()
    });
}

export function removeTokenInfo(callback: Function) {
    chrome.storage.sync.remove('tokenInfo', function () {
        callback()
    });
}

export function setMainLang(lang: string, scene: string) {
    chrome.storage.sync.set({ mainLang: lang })
    if (scene != "init") {
        eventToGoogle({
            name: 'setMainLang',
            params: {
                lang,
                scene
            }
        })
    }

}

export function setSecondLang(lang: string) {
    chrome.storage.sync.set({ secondLang: lang })
    eventToGoogle({
        name: 'setSecondLang',
        params: {
            lang
        }
    })
}

export function openOptionsPage(msg: any) {
    chrome.storage.sync.set({
        optionPageOpenParmas: msg
    }, () => {
        chrome.runtime.openOptionsPage()
        eventToGoogle({
            name: "open_options_page",
            params: msg
        })
    })
}

export function getOptionOpenParmas(): Promise<IAllStorage> {
    return new Promise<IAllStorage>((resolve, _) => {
        chrome.storage.sync.get(['optionPageOpenParmas'], (result: any) => {
            resolve(result)
            // chrome.storage.sync.remove('optionPageOpenParmas')
        })
    })
}

export function onInstall(details: any) {
    let reason = 'unknown'
    if (details.reason === 'install') {
        chrome.tabs.create({ url: "https://www.fishfit.fun/bqy/web/install" })
        chrome.storage.sync.set({ installTime: new Date().valueOf() })
        reason = 'install'
    } else {
        reason = details.reason + '_' + clientVersion
        reason = reason.replace(/\./g, '_')
    }
    eventToGoogle({
        name: reason,
        params: {
            previousVersion: details.previousVersion
        }
    })

    checkLang()
}

export function bgInit() {
    // _mark add conf data
    chrome.storage.sync.get(['bgInit', 'mode', 'mainLang', 'secondLang', 'showProun', 'isTreadWord', 'transEngine', 'keyDownTrans'], (res: any) => {
        const now = new Date().valueOf()
        if (!res.mode) res.mode = 'noset'
        if (res.bgInit) {
            if (now - res.bgInit < 600000) return
            chrome.storage.sync.set({
                bgInit: new Date().valueOf()
            })
            eventToGoogle({
                name: "bg_init",
                params: {
                    transMode: res.mode,
                    mainLang: res.mainLang,
                    secondLang: res.secondLang,
                    showProun: res.showProun,
                    isTreadWord: res.isTreadWord,
                    transEngine: res.transEngine,
                    keyDownTrans: res.keyDownTrans,
                    wrap: `${res.mode}_${res.mainLang}_${res.secondLang}_${res.showProun}_${res.isTreadWord}_${res.transEngine}_${res.keyDownTrans}`
                }
            })
        } else {
            chrome.storage.sync.set({
                bgInit: {
                    lastToAnalytic: new Date().valueOf()
                }
            })
            eventToGoogle({
                name: "bg_init",
                params: {
                    transMode: res.mode,
                    mainLang: res.mainLang,
                    secondLang: res.secondLang,
                    showProun: res.showProun,
                    isTreadWord: res.isTreadWord,
                    transEngine: res.transEngine,
                    keyDownTrans: res.keyDownTrans,
                    wrap: `${res.mode}_${res.mainLang}_${res.secondLang}_${res.showProun}_${res.isTreadWord}_${res.transEngine}_${res.keyDownTrans}`
                }
            })
        }
    })
}

export async function openPDFReader(scene: "option" | "actionMenu" | "openLink", link: string = '') {
    let url = chrome.runtime.getURL("pdf_viewer/web/index.html")
    let openType = "none"
    if (scene === 'actionMenu') {
        link = await getCurrentTabUrl()
    }
    const len = link.length

    if (len > 4) {
        if (scene === 'openLink') {
            url = url + "?url=" + link
        } else if (link.substr(len - 4) === ".pdf") {
            url = url + "?url=" + link
        }
        const s = link.split('://')
        if (s.length > 1) {
            openType = s[0]
        }
    }

    if (scene === 'openLink') {
        window.location.href = url
    } else {
        chrome.tabs.create({ url })
    }

    eventToGoogle({ name: "open_pdf_reader", params: { scene, openType } })
}

function getCurrentTabUrl(): Promise<string> {
    return new Promise<string>((resolve) => {
        chrome.tabs.query({ active: true, lastFocusedWindow: true }, tabs => {
            const url = tabs[0].url;
            if (url) {
                resolve(url)
            } else {
                resolve("")
            }
        });
    })
}

export function checkLang() {
    chrome.storage.sync.get(['mainLang'], (res) => {
        if (!res.mainLang) {
            const [mainLang, secondLang] = setLang()
            eventToGoogle({
                name: 'firstSetLang',
                params: {
                    mainLang,
                    secondLang,
                }
            })
        }

    })
}

function setLang(): string[] {
    let secondLang: string;
    let mainLang: string
    const lang = navigator.language
    if (lang === 'zh-CN' || lang === 'zh') {
        mainLang = 'zh-CN';
    } else if (lang === 'zh-TW' || lang === 'zh-HK') {
        mainLang = 'zh-TW';
    } else {
        //@ts-ignore
        if (languages[lang]) {
            mainLang = lang
        } else {
            mainLang = 'en'
        }
    }
    if (mainLang !== 'en') {
        secondLang = 'en'
    } else {
        secondLang = 'zh-CN'
    }
    chrome.storage.sync.set({ mainLang, secondLang })
    return [mainLang, secondLang]
}