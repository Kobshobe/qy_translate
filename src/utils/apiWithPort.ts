import { applyBDDM, collectResult, reduceCollect, updateMark } from '@/api/api'
import { wrapTranslator } from '@/translator/transWrap'
import { getAudioBase64 } from '@/translator/tts'
import { openOptionsPage } from '@/utils/chromeApi'
import { eventToGoogle } from './analytics'
import { Context } from '@/api/context'

export const apiWrap = {
  translate: async (msg: Context, port: any) => {
    port.postMessage(await wrapTranslator.trans(msg))
  },
  collect: async (c: Context, port:any) => {
    port.postMessage(await collectResult(c))
  },
  reduceCollect: async (c: Context, port: any) => {
    port.postMessage(await reduceCollect(c))
  },
  updateMark: async (c: Context, port: any) => {
    port.postMessage(await updateMark(c))
  },
  tts: async (c: Context, port: any) => {
    port.postMessage(await getAudioBase64(c))
  },
  openOptionsPage: (c: Context, port: any) => {
    openOptionsPage(c.req)
  },
  analytic: async (c: Context, port: any) => {
    eventToGoogle(c.req)
  },
  applyBDDM: async (c: Context, port: any) => {
    port.postMessage(await applyBDDM(c))
  }
}