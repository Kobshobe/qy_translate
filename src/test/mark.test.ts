import { newMarkManager, getMarkHtml } from '../utils/mark'
import {IMarkInfo} from '@/interface/trans'

let marksList = [[3, 5], [10, 14], [18, 22]]
let mark = [4, 7]
let markInfo: IMarkInfo;

test('mark test newMarkInfo', () => {
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.newMarksList).toMatchObject([[3, 7], [10, 14], [18, 22]])
    expect(markInfo.newMarksList).not.toEqual([[3, 7], [10, 14], [18, 22], [0, 0]])
    expect(false).toBe(markInfo.canReduce)

    mark = [5, 7]
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.newMarksList).toEqual([[3, 7], [10, 14], [18, 22]])
    expect(false).toBe(markInfo.canReduce)

    mark = [5, 10]
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.newMarksList).toEqual([[3, 14], [18, 22]])
    expect(false).toBe(markInfo.canReduce)

    mark = [4, 12]
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.newMarksList).toEqual([[3, 14], [18, 22]])
    expect(false).toBe(markInfo.canReduce)

    mark = [8, 15]
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.newMarksList).toEqual([[3, 5], [8, 15], [18, 22]])
    expect(false).toBe(markInfo.canReduce)

    mark = [3, 22]
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.newMarksList).toEqual([[3, 22]])
    expect(false).toBe(markInfo.canReduce)
})

test('mark test canReduce', () => {
    mark = [3, 5]
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.canReduce).toBe(true)
    expect(markInfo.newMarksList).toEqual([[10, 14], [18, 22]])

    mark = [10, 14]
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.canReduce).toBe(true)
    expect(markInfo.newMarksList).toEqual([[3, 5], [18, 22]])

    mark = [11, 13]
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.canReduce).toBe(true)
    expect(markInfo.newMarksList).toEqual([[3, 5], [10, 11], [13, 14], [18, 22]])
})

test('mark test single and empty', () => {
    mark = [3, 5]
    marksList = []
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.canReduce).toBe(false)
    expect(markInfo.newMarksList).toEqual([[3, 5]])

    mark = [3, 5]
    marksList = [[8, 9]]
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.canReduce).toBe(false)
    expect(markInfo.newMarksList).toEqual([[3, 5], [8, 9]])

    mark = [3, 9]
    marksList = [[3, 9]]
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.canReduce).toBe(true)
    expect(markInfo.newMarksList).toEqual([])

    mark = [4, 8]
    marksList = [[3, 9]]
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.canReduce).toBe(true)
    expect(markInfo.newMarksList).toEqual([[3, 4], [8, 9]])

    mark = [3, 5]
    marksList = [[3, 9]]
    markInfo = newMarkManager(mark, marksList)
    expect(markInfo.canReduce).toBe(true)
    expect(markInfo.newMarksList).toEqual([[5, 9]])
})

test('generate markHtml test', () => {
    let markHtml = ''
    let text = `apple pen is attractive`
    const markTabStar = `<span class="mark-text-wsrfhedsoufheqiwrhew">`
    const markTabEnd = `</span>`

    marksList = [[0, 5]]
    markHtml = getMarkHtml(marksList, text)
    expect(markHtml).toBe(`${markTabStar}apple${markTabEnd} pen is attractive`)

    marksList = [[6, 9]]
    markHtml = getMarkHtml(marksList, text)
    expect(markHtml).toBe(`apple ${markTabStar}pen${markTabEnd} is attractive`)

    marksList = [[10, 23]]
    markHtml = getMarkHtml(marksList, text)
    expect(markHtml).toBe(`apple pen ${markTabStar}is attractive${markTabEnd}`)

    marksList = [[0, 5]]
    markHtml = getMarkHtml(marksList, text)
    expect(markHtml).toBe(`${markTabStar}apple${markTabEnd} pen is attractive`)
    marksList = newMarkManager([10, 23], marksList).newMarksList
    markHtml = getMarkHtml(marksList, text)
    expect(markHtml).toBe(`${markTabStar}apple${markTabEnd} pen ${markTabStar}is attractive${markTabEnd}`)

    marksList = [[0, 23]]
    markHtml = getMarkHtml(marksList, text)
    expect(markHtml).toBe(`${markTabStar}apple pen is attractive${markTabEnd}`)
    // 取消标记
    marksList = newMarkManager([6, 9], marksList).newMarksList
    markHtml = getMarkHtml(marksList, text)
    expect(markHtml).toBe(`${markTabStar}apple ${markTabEnd}pen${markTabStar} is attractive${markTabEnd}`)
    expect(marksList).toEqual([[0,6],[9, 23]])
})
