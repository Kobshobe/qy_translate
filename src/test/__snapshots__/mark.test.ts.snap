// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mark test 1`] = `
Array [
  Array [
    10,
    14,
  ],
  Array [
    18,
    22,
  ],
  Array [
    3,
    7,
  ],
]
`;
